Make it easier to build the dataset pipeline in Dataset C++ tests

feihugis opened the pull request:
This PR aims to simplify the code of making the dataset pipeline in Dataset C++ tests using the following style:
```C++
auto* source_dataset_params = new RangeDatasetParams(...);
auto* target_dataset_params = source_dataset_params->Batch(...)->Map(...)->...->...;
// Handles all the allocated dataset_params (source_dataset_params, batch_dataset_params, map_dataset_params, ... , target_dataset_params)
core::ScopedUnref scoped_unref(target_dataset_params); 
Tensor target_dataset;
MakeDatasetTensor(target_dataset_params, &target_dataset);
...
...

```
One example is as below
```C++
// The ownership is transferred to the caller.
MapDatasetParams* MapDatasetParams() {
  auto* range_dataset_params = new RangeDatasetParams(10, 0, -3);
  return range_dataset_params
      ->Batch(
          /*batch_size=*/2,
          /*drop_remainder=*/false,
          /*parallel_copy=*/true,
          /*output_dtypes=*/{DT_INT64},
          /*output_shapes=*/{PartialTensorShape({2})},
          /*node_name=*/"batch_node")
      ->Map(
          /*other_arguments=*/{},
          /*func=*/
          FunctionDefHelper::FunctionRef("XAddX", {{"T", DT_INT64}}),
          /*func_lib=*/{test::function::XAddX()},
          /*type_arguments=*/{},
          /*output_dtypes=*/{DT_INT64},
          /*output_shapes=*/{PartialTensorShape({1})},
          /*use_inter_op_parallelism=*/true,
          /*preserve_cardinality=*/false,
          /*node_name=*/kNodeName);
}
```

feihugis committed:
Make it easier to chain multiple datasets in Dataset C++ tests

feihugis commented:
@jsimsa The first commit (https://github.com/tensorflow/tensorflow/pull/31635/commits/79b55fd17b055c4a17737de87db6e0667685e2e5) is as same as #31592, which you have reviewed and is pending for merge. You can just review the second commit (https://github.com/tensorflow/tensorflow/pull/31635/commits/08ebb28d857e425134485487e3c65aeb38f0631c).

feihugis commented on tensorflow/core/kernels/data/dataset_test_params.h:
```diff
@@ -0,0 +1,208 @@
+
+  ~UnaryDatasetParams() {
+    if (input_dataset_params) {
+      input_dataset_params->Unref();
```
Handles the release of input dataset params in the data pipeline.

jsimsa replied:
seeing as the ownership of this params object is never shared, I don't think you need refcounting or smart pointer and simply passing / moving the object values around will suffice

feihugis commented on tensorflow/core/kernels/data/dataset_test_params.h:
```diff
@@ -0,0 +1,208 @@
+    }
+    if (input_dataset_params_1) {
+      input_dataset_params_1->Unref();
+    }
```
Handles the release of input dataset params in the data pipeline.

feihugis commented on tensorflow/core/kernels/data/dataset_test_params.h:
```diff
@@ -0,0 +1,208 @@
+  static bool IsDatasetTensor(const Tensor& tensor);
+
+  // Creates a BatchDatasetParams. The ownership is transferred to the caller.
+  BatchDatasetParams* Batch(int64 batch_size, bool drop_remainder,
```
The ownership is transferred to the caller. That means the caller just needs to call `batch_dataset_params->UnRef()` or `core::ScopedUnref scoped_unref(batch_dataset_params)` to handle the deallocation for the whole pipeline.

feihugis replied:
For example:
```C++
auto* range_dataset_params = new RangeDatasetParams(10, 0, -3);
auto* map_dataset_params = range_dataset_params->Batch(...)->Map(...);
core::ScopedUnref scoped_unref(map_dataset_params);
```
The allocated `range_dataset_params`, `batch_dataset_params`, and `map_dataset_params` will be automatically released.

jsimsa replied:
Having these as member methods only works for unary transformations. However, for dataset sources (e.g. `range` or `random`) or binary (e.g. `concatenate`) or n-nary transformation (e.g. `zip`) this will not work.

Instead of a member method, there should be a static factory that takes the input dataset parameters as a parameter:

```
static DatasetParams Range(...);
static DatasetParams Batch(DatasetParams params, ...);
static DatasetParams Zip(std::vector<DatasetParams> params, ...);
```

and then use it as follows:

```
auto params = DatasetParams::CreateMap(std::move(DatasetParams::CreateBatch(std::move(DatasetParams::CreateRange(10, 0, -3)), ...)), ...);
```

feihugis replied:
@jsimsa Thanks for the prompt review! I have a few questions here.

>Instead of a member method, there should be a static factory that takes the input dataset parameters as a parameter

One potential issue for the static factory is that different static methods will have different input and return types as different DatasetParams have different attributes. For example:
```C++
static RangeDatasetParams Range(...);
static BatchDatasetParams Batch(T params, ...);
static ZipDatasetParams Zip(std::vector<DatasetParams> params, ...);
```
>`auto params = DatasetParams::Map(std::move(DatasetParams::Batch(std::move(DatasetParams::Range(10, 0, -3)), ...)), ...);`

The different types of input and return DatasetParams in each static method make it difficult to chain multiple dataset parameters together.

jsimsa replied:
Why can't they all return `DatasetParams` which the derived classes inherit from?

feihugis replied:
I may misunderstand something. Here is an example:
```C++
static DatasetParams Range(int64 start, int64 stop, int64 step) {
  return RangeDatasetParams(start, stop, step);
}

DatasetParams range_dataset_params = Range(start, stop, step);
```
during the conversion, the values of `range_dataset_params.start`, `range_dataset_params.stop`, and `range_dataset_params.step` will be discarded.

jsimsa commented on tensorflow/core/framework/function_testlib.cc:
```diff
@@ -585,6 +585,76 @@ FunctionDef MakeRangeDataset() {
+         {"preserve_cardinality", "$preserve_cardinality"}}}});
+}
+
+FunctionDef MakeMapDatasetWithoutOtherArgs() {
```
there should be a single function with a boolean argument that controls `other_arguments` will be included

feihugis replied:
Yeah, will change that.

jsimsa commented on tensorflow/core/kernels/data/dataset_test_params.h:
```diff
@@ -0,0 +1,208 @@
+class BatchDatasetParams;
+class MapDatasetParams;
+
+class DatasetParams : public core::RefCounted {
```
prefer using smart pointer over RefCounted ... but I don't think you need either ... you should be able to create static factories that return params object values and move these into subsequent factory invocations.

jsimsa replied:
see my comment below

jsimsa reviewed (changes requested):
Did a pass and left some comments.

feihugis committed:
Address the comments

feihugis commented:
@jsimsa Thanks for your suggestion! Now `RefCounted` is replaced by the smart pointer and the static factory methods are added as below. Could you please have a look at the changes (https://github.com/tensorflow/tensorflow/pull/31635/commits/50624183a1e6c92546c268687afc6c6c817f8777) when you get a chance?
```C++
static std::shared_ptr<BatchDatasetParams> Batch(
    const std::shared_ptr<DatasetParams>& input_dataset_params, ...);

static std::shared_ptr<MapDatasetParams> Map(
    const std::shared_ptr<DatasetParams>& input_dataset_params, ...) ;

static std::shared_ptr<RangeDatasetParams> Range(int64 start, int64 stop, int64 step);
```

feihugis commented on tensorflow/core/kernels/data/dataset_test_params.h:
```diff
@@ -0,0 +1,184 @@
+  // Used to store all the input dataset parameters and the dataset tensors
+  // generated from the parameters.
+  std::vector<std::pair<std::shared_ptr<DatasetParams>, Tensor>>
+      input_dataset_params_group;
```
`input_dataset_params_group` is used to handle n-nary transformation.

mellanox-github commented:
Can one of the admins verify this patch?

feihugis committed:
Add DatasetParamsBuilder to build the dataset pipeline

jsimsa reviewed (commented):
Thanks @feihugis did a pass and left a bunch of comments.

feihugis committed:
Update the constructor without using smart pointer and make data members be private

feihugis commented on tensorflow/core/kernels/data/dataset_test_params.h:
```diff
@@ -139,102 +195,14 @@ class RangeDatasetParams : public DatasetParams {
-  Tensor step;
-};
-
-static std::shared_ptr<BatchDatasetParams> Batch(
```
With the updated constructors of these DatasetParams class, the static factories are not needed.

feihugis commented on tensorflow/core/kernels/data/batch_dataset_op_test.cc:
```diff
@@ -236,33 +212,43 @@ ITERATOR_GET_NEXT_TEST_P(BatchDatasetOpTest, BatchDatasetParams,
 }
 
-std::vector<DatasetOutputShapesTestCase<BatchDatasetParams>>
+std::vector<DatasetOutputShapesTestCase<std::shared_ptr<BatchDatasetParams>>>
```
`shared_ptr` is removed now.

feihugis commented on tensorflow/core/kernels/data/dataset_test_params.cc:
```diff
@@ -0,0 +1,135 @@
+}
+
+BatchDatasetParams::BatchDatasetParams(
+    std::shared_ptr<DatasetParams> input_dataset_params, int64 batch_size,
```
After adding the implementation of `DatasetParams::MakeInputs`, the errors disappeared. But some members of the derived classes lose their value during the copy construction. One example can be found [here](https://coliru.stacked-crooked.com/view?id=ed75556c39348c9a) (the layer area should be same with the rectangle area).

The constructor is changed to be as below without using the smart pointer:
```C++
template <typename T>
  BatchDatasetParams(T input_dataset_params, int64 batch_size,
                     bool drop_remainder, bool parallel_copy,
                     DataTypeVector output_dtypes,
                     std::vector<PartialTensorShape> output_shapes,
                     string node_name)
```
Please feel free to let me know if you have any suggestion.

feihugis commented:
@jsimsa Thanks for your review! The smart pointers are removed and all the members in DatasetParams are changed to be private. Could you please have a loot at the changes (https://github.com/tensorflow/tensorflow/pull/31635/commits/55afc02a147669a25e462abfafa684b9103476fa) when you get a chance?

jsimsa commented on tensorflow/core/framework/function_testlib.h:
```diff
@@ -40,6 +40,14 @@ class Attrs {
     }
   }
 
+  Attrs(
```
why is this needed?

feihugis replied:
Some places like [here](https://github.com/tensorflow/tensorflow/pull/31635/files#diff-ba4d21ca0c437c902fd8c791e06ba44dR308) need it to convert [AttributeVector](https://github.com/tensorflow/tensorflow/pull/31635/files#diff-a4a3b8bfea4ac6f0185c2a132dadbb8bR32) to `Attrs`.

feihugis replied:
`AttributeVector` can not be replaced by `Attrs` as it is required by `test::function::NDef` like [here](https://github.com/tensorflow/tensorflow/pull/31635/files#diff-4085eb633f02b68967febffed95bc6edR41).

jsimsa replied:
ack

jsimsa commented on tensorflow/core/kernels/data/batch_dataset_op_test.cc:
```diff
@@ -109,96 +58,106 @@ class BatchDatasetOpTest : public DatasetOpsTestBaseV2<BatchDatasetParams> {
-          /*output_shapes=*/{PartialTensorShape({-1})},
-          /*node_name=*/kNodeName};
+  auto range_dataset_params = RangeDatasetParams(0, 10, 1);
+  auto batch_dataset_params_0 =
```
no need to create a variable for this one, you can just return it

feihugis replied:
Yeah, changed it.

jsimsa commented on tensorflow/core/kernels/data/dataset_test_params.h:
```diff
@@ -0,0 +1,211 @@
+  DatasetParamsType type_;
+};
+
+class BatchDatasetParams : public DatasetParams {
```
I feel like this should be in `batch_dataset_op_test.h`, inlining the implementation of `MakeInputs` and `MakeAttributes` so that the header can be included from test .cc files that wish to use it

Putting all params objects for all dataset op kernels into this file does not scale.

feihugis replied:
Great suggestion! Will change it as you suggest.

jsimsa commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -293,22 +292,61 @@ Status DatasetOpsTestBase::MakeRangeDataset(
 }
 
-// Create a `RangeDataset` dataset as a variant tensor.
 Status DatasetOpsTestBase::MakeRangeDataset(
```
similar to my comment about dataset_params, this should go into `range_dataset_op_test.h` ... putting all test dataset into this files does not scale

feihugis replied:
Yeah, I'm also planning to change this. After `DatasetParams` is updated with `MakeInputs()` and `MakeAttributes()`, we do not need to create a similar function for each dataset. A more general function can be used to replace them as below. 
```C++
Status DatasetOpsTestBase::MakeDatasetTensor(
      DatasetParams& dataset_params, 
      const FunctionDef& make_deataset_func, 
      Tensor* dataset_tensor);
```

feihugis committed:
Split different types of DatasetParams to individual files and Refactor class DatasetOpsTestBaseV2

1. Each specific type of DatasetParams is defined in different files;

2. Refactor the class `DatasetOpsTestBaseV2`:
   a. the derived class does not need to override `Initialize()` and `MakeDatasetOpKernel`.
   b. get rid of the template type.
   c. update `MakeDatasetTensor()` to avoid the functions for each dataset.

feihugis commented:
@jsimsa Thanks for your suggestions! The comments are addressed here(https://github.com/tensorflow/tensorflow/pull/31635/commits/7003529c6e7bdbc5ad6c493fe755ec8d2a88e44f). Also, clean up `DatasetOpsTestBaseV2` a little bit so that the derived test classes do not need to override the functions `Initialize()` and `MakeDatasetOpKernel`. Could you please have another look when you have time?

rachellim commented on tensorflow/core/kernels/data/batch_dataset_op_test.cc:
```diff
@@ -11,194 +11,119 @@ limitations under the License.
-          /*output_shapes=*/{PartialTensorShape({4})},
-          /*node_name=*/kNodeName};
+  auto range_dataset_params = RangeDatasetParams(0, 12, 1);
+  return BatchDatasetParams(std::move(range_dataset_params),
```
```suggestion
  return BatchDatasetParams(RangeDatasetParams(0, 12, 1),
```

feihugis replied:
If my understanding is correct, `std::move(range_dataset_params)` is used to avoid a copy here.

rachellim replied:
I believe the compiler will know to elide the copy for the temporary RangeDatasetParams object in this case -- https://en.cppreference.com/w/cpp/language/copy_elision

feihugis replied:
Yes, you are right. Will change it.

rachellim commented on tensorflow/core/kernels/data/batch_dataset_op_test.h:
```diff
@@ -0,0 +1,86 @@
+namespace tensorflow {
+namespace data {
+
+class BatchDatasetParams : public DatasetParams {
```
I'm in favor of having a dataset test utils file / namespace where these common classes are defined, instead of having a header file for batch dataset test specifically. It makes the test harder to read. Furthermore, you already have a bunch of utility-like files `dataset_test_{base | params}.{h | cc}`. I'd recommend coalescing all that into a single pair of `.h` and `.cc` files.

 I'll defer to @jsimsa on this though, if he thinks otherwise.

feihugis replied:
There are two kinds of `DatasetParams`: 
1) the first one (e.g. `RangeDatasetParams`) will be used as the input dataset by other dataset op tests; 
2) the second kind of DatasetParams (e.g. `ReduceDatasetParams`) will be only used by itself. 

For the first one, we need to define it in some files (e.g. `dataset_test_params.h`) that can be imported by other tests; for the second one, we can define the dataset params class in its own test file (e.g. `reduce_dataset_op_test.cc`) as it will not be used by other tests. It may solve the scale issue that @jsimsa mentioned before.

`dataset_test_base.h` already has more than 600 lines of code, so I did not combine `dataset_test_base.{h | cc}` and `dataset_test_params.{h | cc}` together. If the length of the file is ok,  I will merge them into a single pair of `.h` and `.cc` file once @jsimsa and @rachellim make the final decision. 

Adding a namespace for the tests (e.g. `tensorflow::data::testing`) sounds great to me.

jsimsa replied:
@feihugis if we only expect a small number of dataset parameter definitions to be shared, then @rachellim's suggestion makes sense.

I would not worry about the size of the `dataset_test_base` files as long as they are well organized and documented.

In summary, I support:
1) coalescing all test related functionality in `dataset_test_base.{h, cc}` putting shared dataset parameter definitions there and inlining all dataset parametr definitions in the `*_test.cc` files.

feihugis replied:
Great, will coalesce these `_params.h` files into  the `dataset_test_base` files.

rachellim commented on tensorflow/core/kernels/data/dataset_test_base.cc:
```diff
@@ -698,5 +678,100 @@ Status DatasetOpsTestBase::CheckIteratorSaveAndRestore(
   return Status::OK();
 }
 
+Status DatasetOpsTestBaseV2::Initialize(DatasetParams& dataset_params) {
```
While we're refactoring the tests, why do we have `DatasetOpsTestBase` and `DatasetOpsTestBaseV2`? Can you add comments that explain the distinction?

feihugis replied:
Before this PR, `DatasetOpsTestBaseV2` will break the existing tests based on `DatasetOpsTestBase`. The changes in this PR will enable us to merge `DatasetOpsTestBaseV2` and `DatasetOpsTestBase`. Will do it once this PR is finalized.

rachellim reviewed (commented):
I haven't looked extensively at all of the changes here, but some comments to start.

feihugis commented on tensorflow/core/framework/function_testlib.cc:
```diff
@@ -585,6 +585,58 @@ FunctionDef MakeRangeDataset() {
+                  ? std::vector<string>{"input_dataset: variant",
+                                        "other_arguments: Targuments"}
+                  : std::vector<string>{"input_dataset: variant"};
+  auto input_placeholders =
```
This name is not good. How about `node_args`?

rachellim replied:
What about just `inputs`?

feihugis replied:
Yeah, sounds good. Will change it.

feihugis commented on tensorflow/core/framework/function_testlib.cc:
```diff
@@ -585,6 +585,58 @@ FunctionDef MakeRangeDataset() {
 
+FunctionDef MakeBatchDataset() {
+  return FDH::Define(
+      // Name
```
Thanks for the link!

feihugis committed:
Fix the style issues
1. spacing format
2. merge *_params.h into dataset_test_base
3. function argument comment format

feihugis commented:
Thanks @rachellim for your review and help! Most of the comments are addressed here (https://github.com/tensorflow/tensorflow/pull/31635/commits/015b45e95dac3047060f698b959083a4dcc6e2b4) except this one (https://github.com/tensorflow/tensorflow/pull/31635#discussion_r320406144). Will do it once we get the answer from @jsimsa.

feihugis commented:
@rachellim All the comments are addressed now (https://github.com/tensorflow/tensorflow/pull/31635/commits/321e8cdf098d6a5b5645a9e3bbff0e0f6f223caf). Could you please take another look when you get a chance?

feihugis commented:
> I'd suggest making everything consistent to _have_ a space after the colon, not removing the space from everything.

Gotcha. Will change them

rachellim commented:
Thanks for your changes!

feihugis committed:
Address the comments

feihugis commented:
Thanks for the very prompt review, @rachellim! This commit (https://github.com/tensorflow/tensorflow/pull/31635/commits/2bf79a761656e3c64fbaeead39e63e41fd492255) is submitted to address the comments. Please take a look!

feihugis committed:
Refactor SamplingDatasetOpTest and MapAndBatchDatasetOpTest

feihugis commented:
The internal checks failed. Could you help check the logs and paste them here?

jsimsa commented:
internal tests are still running

feihugis commented:
@jsimsa The internal checks are in the failed status. Could you please help check the logs?

jsimsa commented:
```
./third_party/tensorflow/core/kernels/data/dataset_test_base.h:132:3: error: 'tensorflow::data::DatasetParams' has virtual functions but non-virtual destructor [-Werror,-Wnon-virtual-dtor]
  ~DatasetParams() {}
  ^
./third_party/tensorflow/core/kernels/data/dataset_test_base.h:184:7: error: 'tensorflow::data::RangeDatasetParams' has virtual functions but non-virtual destructor [-Werror,-Wnon-virtual-dtor]
class RangeDatasetParams : public DatasetParams {
      ^
./third_party/tensorflow/core/kernels/data/dataset_test_base.h:208:7: error: 'tensorflow::data::BatchDatasetParams' has virtual functions but non-virtual destructor [-Werror,-Wnon-virtual-dtor]
class BatchDatasetParams : public DatasetParams {
      ^
./third_party/tensorflow/core/kernels/data/dataset_test_base.h:245:7: error: 'tensorflow::data::MapDatasetParams' has virtual functions but non-virtual destructor [-Werror,-Wnon-virtual-dtor]
class MapDatasetParams : public DatasetParams {
```

feihugis committed:
Add the virtual destructor

feihugis commented:
@jsimsa The virtual deconstructor is added [here](https://github.com/tensorflow/tensorflow/pull/31635/commits/b0040a837c27bbd81e5f0033b4cd493a8f2513a9) and also update some places to use the named constant. Could you please take another look when you get a chance?

feihugis commented on tensorflow/core/kernels/data/dataset_test_base.h:
```diff
@@ -129,7 +129,7 @@ class DatasetParams {
                 DatasetParamsType type);
 
-  ~DatasetParams() {}
+  virtual ~DatasetParams() {}
```
Add the virtual destructor for the base class, but did not add the destructor for the derived class in order to make the code concise (also the destructor does not do anything here). But please feel free to let me know if it is a good practice to add the destructor for the derived classes.

feihugis commented:
@rthadur An error happed while migrating the change. Could you help retrigger the internal checks? Thanks!

rthadur commented:
done
